utils/python/CIME/XML/generic_xml.py

```diff
@@ -63,55 +58,72 @@ def write(self, outfile=None):
             outfile = self.filename
 
         logger.debug("write: " + outfile)
-        xmlstr = ET.tostring(self.root)
-        doc = minidom.parseString(xmlstr)
-        with open(outfile,'w') as xmlout:
-            doc.writexml(xmlout,addindent='  ')
+        try:
+            xmlstr = ET.tostring(self.root)
+        except:
+            ET.dump(self.root)
+            expect(False, "Could not write file %s, xml formatting error"%self.filename)
+
+        # xmllint provides a better format option for the output file
+        xmllint = which("xmllint")
+        if xmllint is not None:
+            run_cmd("%s --format --output %s -"%(xmllint,outfile), input_str=xmlstr)
+        else:
+            doc = minidom.parseString(xmlstr)
+            with open(outfile,'w') as xmlout:
+                doc.writexml(xmlout,addindent='  ')
 
-    def get_node(self, nodename, attributes=None, root=None):
+
+
+    def get_node(self, nodename, attributes=None, root=None, xpath=None):
         """
         Get an xml element matching nodename with optional attributes.
 
         Error unless exactly one match.
         """
-        nodes = self.get_nodes(nodename, attributes=attributes, root=root)
+        nodes = self.get_nodes(nodename, attributes=attributes, root=root, xpath=xpath)
         expect(len(nodes) == 1, "Incorrect number of matches, %d, for nodename '%s' and attrs '%s' in file '%s'" %
                (len(nodes), nodename, attributes, self.filename))
         return nodes[0]
 
-    def get_optional_node(self, nodename, attributes=None, root=None):
+    def get_optional_node(self, nodename, attributes=None, root=None, xpath=None):
         """
         Get an xml element matching nodename with optional attributes.
 
         Return None if no match.
         """
-        nodes = self.get_nodes(nodename, attributes=attributes, root=root)
+        nodes = self.get_nodes(nodename, attributes=attributes, root=root, xpath=xpath)
 
         expect(len(nodes) <= 1, "Multiple matches for nodename '%s' and attrs '%s' in file '%s'" %
                (nodename, attributes, self.filename))
         return nodes[0] if nodes else None
 
-    def get_nodes(self, nodename, attributes=None, root=None):
+    def get_nodes(self, nodename, attributes=None, root=None, xpath=None):
         if root is None:
             root = self.root
         nodes = []
-        xpath = ".//"+nodename
+        expect(attributes is None or xpath is None,
+               " Arguments attributes and xpath are exclusive")
+        if xpath is None:
+            xpath = ".//"+nodename
         if attributes is not None:
             # xml.etree has limited support for xpath and does not allow more than
             # one attribute in an xpath query so we query seperately for each attribute
             # and create a result with the intersection of those lists
             for key, value in attributes.iteritems():
```

Comment:
This is the input argument attributes which is used to create an xpath to search the xml object.  I found that in some cases the xpath was more complicated than that so I choose to optionally pass in the xpath argument directly instead of using the attributes argument.   I think that we could make this two different functions instead.